lxd/storage/drivers/drivers_lvm_volumes.go

```diff
@@ -0,0 +1,732 @@
+	}
+
+	if volTargetArgs.MigrationType.FSType != migration.MigrationFSType_RSYNC {
+		return fmt.Errorf("Migration type not supported")
+	}
+
+	return genericCreateVolumeFromMigration(d, nil, vol, conn, volTargetArgs, preFiller, op)
+}
+
+// CreateVolumeFromCopy provides same-pool volume copying functionality.
+func (d *lvm) CreateVolumeFromCopy(vol, srcVol Volume, copySnapshots bool, op *operations.Operation) error {
+	var err error
+	var srcSnapshots []Volume
+
+	if vol.contentType != ContentTypeFS || srcVol.contentType != ContentTypeFS {
+		return fmt.Errorf("Content type not supported")
+	}
+
+	if copySnapshots && !srcVol.IsSnapshot() {
+		// Get the list of snapshots from the source.
+		srcSnapshots, err = srcVol.Snapshots(op)
+		if err != nil {
+			return err
+		}
+	}
+
+	// We can use optimised copying when the pool is backed by an LVM thinpool.
+	if d.usesThinpool() {
+		return d.copyThinpoolVolume(vol, srcVol, srcSnapshots, false)
+	}
+
+	// Otherwise run the generic copy.
+	return genericCopyVolume(d, nil, vol, srcVol, srcSnapshots, false, op)
+}
+
+// RefreshVolume provides same-pool volume and specific snapshots syncing functionality.
+func (d *lvm) RefreshVolume(vol, srcVol Volume, srcSnapshots []Volume, op *operations.Operation) error {
+	if vol.contentType != ContentTypeFS || srcVol.contentType != ContentTypeFS {
+		return fmt.Errorf("Content type not supported")
+	}
+
+	// We can use optimised copying when the pool is backed by an LVM thinpool.
+	if d.usesThinpool() {
+		return d.copyThinpoolVolume(vol, srcVol, srcSnapshots, true)
+	}
+
+	// Otherwise run the generic copy.
+	return genericCopyVolume(d, nil, vol, srcVol, srcSnapshots, true, op)
+}
+
+// VolumeSnapshots returns a list of snapshots for the volume.
+func (d *lvm) VolumeSnapshots(vol Volume, op *operations.Operation) ([]string, error) {
+	// We use the vfsVolumeSnapshots rather than inspecting the logical volumes themselves because the origin
+	// property of an LVM snapshot can be removed/changed when restoring snapshots, such that they are no
+	// marked as origin of the parent volume.
+	return d.vfsVolumeSnapshots(vol, op)
+}
+
+// UpdateVolume applies config changes to the volume.
+func (d *lvm) UpdateVolume(vol Volume, changedConfig map[string]string) error {
+	if vol.contentType != ContentTypeFS {
+		return fmt.Errorf("Content type not supported")
+	}
+
+	if _, changed := changedConfig["size"]; changed {
+		err := d.SetVolumeQuota(vol, changedConfig["size"], nil)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// RenameVolume renames a volume and its snapshots.
+func (d *lvm) RenameVolume(vol Volume, newVolName string, op *operations.Operation) error {
+	volDevPath := d.lvmDevPath(d.config["lvm.vg_name"], vol.volType, vol.contentType, vol.name)
+	lvExists, err := d.logicalVolumeExists(volDevPath)
+	if err != nil {
+		return err
+	}
+
+	if !lvExists {
+		return fmt.Errorf("LVM logical volume doesn't exist")
+	}
```

Comment:
Why the early check? We should never be called if it doesn't exist, right?
If so, then we can rely on the rename error for the unlikely event where it indeed does not exist.

Reply:
I've removed this.